parser.js

```diff
@@ -0,0 +1,119 @@
+var CompileError = require("./compileerror");
+var solc = require("solc");
+
+// Clean up after solc.
+var listeners = process.listeners("uncaughtException");
+var solc_listener = listeners[listeners.length - 1];
+
+if (solc_listener) {
+  process.removeListener("uncaughtException", solc_listener);
+}
+
+module.exports = {
+  parse: function(body, fileName) {
+    // Here, we want a valid AST even if imports don't exist. The way to
+    // get around that is to tell the compiler, as they happen, that we 
+    // have source for them (an empty file).
+
+    var fileName = fileName || "ParsedContract.sol";
+
+    var solcStandardInput = {
+      language: "Solidity",
+      sources: {
+        [fileName]: {
+          content: body
+        }
+      },
+      settings: {
+        outputSelection: {
+          "Contract.sol": {
+            "*": ["ast"] 
+          }
+        }
+      }
+    };
+
+    var output = solc.compileStandard(JSON.stringify(solcStandardInput), function(file_path) {
+      // Tell the compiler we have source code for the dependency
+      return {contents: "pragma solidity ^0.4.0;"};
+    });
+
+    output = JSON.parse(output);
+
+    if (output.errors) {
+      throw new CompileError(output.errors[0].formattedMessage);
+    }
+
+    return {
+      contracts: Object.keys(output.contracts[fileName]),
+      ast: output.sources[fileName].ast
+    };
+  },
+
+  // This needs to be fast! It is fast (as of this writing). Keep it fast!
+  parseImports: function(body) {
+    var self = this;
+    
+    // So we don't have to maintain a separate parser, we'll get all the imports
+    // in a file by sending the file to solc and evaluating the error messages 
+    // to see what import statements couldn't be resolved. To prevent full-on
+    // compilation when a file has no import statements, we inject an import
+    // statement right on the end; just to ensure it will error and we can parse
+    // the imports speedily without doing extra work.
+
+    // Helper to detect import errors with an easy regex. 
+    var importErrorKey = "TRUFFLE_IMPORT";
+
+    // Inject failing import.
+    var failingImportFileName = "__Truffle__NotFound.sol";
+    
+    body = body + "\n\nimport '" + failingImportFileName + "';\n";
+
+    var solcStandardInput = {
+      language: "Solidity",
+      sources: {
+        "Contract.sol": {
+          content: body
+        }
+      },
+      settings: {
+        outputSelection: {
+          "Contract.sol": {
+            "*": [] // We don't need any output.
+          }
+        }
+      }
+    };
+
+    var output = solc.compileStandard(JSON.stringify(solcStandardInput), function() {
+      // The existence of this function ensures we get a parsable error message.
+      // TRUFFLE_IMPORT lets us make a nice, detectable error. 
```

Comment:
otherwise is it a string or something?

Reply:
This just adds the word "TRUFFLE_IMPORT" to all error messages that that resulted from missing dependencies. Not necessarily needed, but makes detection much easier (i.e., a simple indexOf).